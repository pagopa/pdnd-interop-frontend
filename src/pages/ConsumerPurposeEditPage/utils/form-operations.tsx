import {
  CheckboxGroup,
  RadioGroup,
  Select,
  TextField,
} from '@/components/shared/ReactHookFormInputs'
import type { InputOption } from '@/types/common.types'
import React from 'react'
import identity from 'lodash/identity'
import type {
  Dependency,
  DynamicFormOperations,
  Questions,
  QuestionV1,
  QuestionV2,
} from '../types/risk-analysis.types'
import { RiskAnalysisSwitch } from '../components/PurposeEditStep2RiskAnalysis/RiskAnalysisSwitch'

export const dynamicFormOperationsVersions: DynamicFormOperations = {
  '1.0': {
    getUpdatedQuestions: (values, riskAnalysis) => {
      const updatedQuestions = riskAnalysis.questions.reduce((acc, next) => {
        // We are sure we can only queue the questions that come after the current one
        // because the dependencies are always on the questions that come before.
        // E.g. question 3 can only have question 1 and/or 2 as a dependency,
        // not question 4 and subsequent
        const queuedQuestions = Object.values(acc).map((d) => d.id)

        // Check the dependencies of this questions
        const satisfiesAllDependencies = next.dependencies.every((d) => {
          // If the question is a dependency
          const isInArray = queuedQuestions.includes(d.id)
          const id = d.id as string

          // and if the field currently has the same value as stated in the dependency
          const hasSameValue = Array.isArray(values[id])
            ? (values[id] as Array<unknown>).includes(d.value)
            : values[id] === d.value

          return isInArray && hasSameValue
        })

        // Then it needs to be added as a new question to ask
        if (satisfiesAllDependencies) {
          acc[next.id] = next
        }

        return acc
      }, {} as Questions)
      return updatedQuestions
    },

    buildForm: (questions, _, lang, t) => {
      const questionKeys = Object.keys(questions)

      const formComponents = questionKeys.map((id, i) => {
        const { type, label, options, infoLabel } = questions[id] as QuestionV1

        const sx = questionKeys.length - 1 === i ? { mb: 0 } : {}
        const inputOptions = options ? options.map((o) => ({ ...o, label: o.label[lang] })) : []
        const commonProps = {
          key: id,
          name: id,
          label: label[lang],
          infoLabel: infoLabel ? infoLabel[lang] : undefined,
          sx,
        }

        switch (type) {
          case 'text':
            return <TextField {...commonProps} rules={{ required: true }} />

          case 'select-one':
            return (
              <Select
                {...commonProps}
                options={inputOptions}
                emptyLabel={t('edit.step2.emptyLabel')}
                rules={{ required: true }}
              />
            )

          case 'checkbox':
            const errorMessage =
              options && options?.length > 1
                ? t('edit.step2.multiCheckboxField.validation.mixed.required')
                : t('edit.step2.singleCheckboxField.validation.mixed.required')
            return (
              <CheckboxGroup
                {...commonProps}
                options={inputOptions}
                rules={{
                  validate: (value) =>
                    (typeof value !== 'undefined' && value.length > 0) || errorMessage,
                }}
              />
            )

          case 'radio':
            return <RadioGroup {...commonProps} options={inputOptions} rules={{ required: true }} />
        }
      })

      return { formComponents: formComponents, isSubmitBtnDisabled: false }
    },
  },
  '2.0': {
    getUpdatedQuestions: (values, riskAnalysis) => {
      const questions = riskAnalysis.questions as Array<QuestionV2>

      // Filters all the questions that not satisfies the dependency inside the "dependencies" property
      const updatedQuestions = questions.filter(({ dependencies }) => {
        function satisfiesDependency(dep: QuestionV2['dependencies'][0]) {
          if (Array.isArray(values[dep.id])) {
            return (values[dep.id] as Array<unknown>).includes(dep.value)
          }
          if (Array.isArray(dep.value)) {
            return dep.value.includes(values[dep.id])
          }
          return values[dep.id] === dep.value
        }
        // If length of dependencies array is 0 the "every" method returns true
        return dependencies.every(satisfiesDependency)
      })

      // Array<Question> -> Record<"id", Question>
      return updatedQuestions.reduce((acc, next) => ({ ...acc, [next.id]: next }), {} as Questions)
    },

    buildForm: (questions, formMethods, lang, t) => {
      const values = formMethods.getValues()

      const formComponents: Array<React.ReactNode> = []

      const questionIds = Object.keys(questions)

      type QuestionOption = Exclude<QuestionV2['options'], undefined>[0]

      function checkOptionDependency(dependency: Dependency) {
        const currentDepValue = values[dependency.id]

        const hasDependency =
          currentDepValue === dependency.value ||
          (Array.isArray(currentDepValue) &&
            (currentDepValue as string[]).includes(dependency.value as string))

        return hasDependency
      }

      function parseOption(option: QuestionOption, { hideOption, id, defaultValue }: QuestionV2) {
        // if the key "hideOption" is present in the question object and the conditions are satisfied
        // the option will be not added to the array of options
        const shouldHideOption =
          hideOption &&
          hideOption[option.value] &&
          hideOption[option.value].some(checkOptionDependency)

        if (!shouldHideOption) {
          return { value: option.value, label: option.label[lang] }
        }

        if ((values[id] as string[])?.includes(option.value)) {
          formMethods.setValue(id, defaultValue)
        }
      }

      function buildFormQuestionComponents(
        { id, type, ...question }: QuestionV2,
        inputOptions: Array<InputOption>,
        isLast: boolean
      ) {
        const questionComponents: Array<React.ReactNode> = []

        const maxLength = question?.validation?.maxLength
        const isRequired = question.required
        const isMultipleChoice = question.dataType === 'multi'

        let label = question.label[lang]
        let infoLabel = question.infoLabel && question.infoLabel[lang]

        if (maxLength) {
          const maxLengthLabel = t('edit.step2.validation.maxLength', { num: maxLength })
          if (infoLabel) {
            infoLabel += `. ${maxLengthLabel}`
          } else {
            infoLabel = maxLengthLabel
          }
        }

        const labelValidation: Array<string> = []

        if (isRequired) {
          labelValidation.push(t('edit.step2.validation.required'))
        }

        if (isMultipleChoice) {
          labelValidation.push(t('edit.step2.validation.multipleChoice'))
        }

        if (labelValidation.length > 0) {
          label += ` (${labelValidation.join(', ')})`
        }

        const sx = isLast ? { mb: 0 } : {}
        const commonProps = { key: id, name: id, label, infoLabel, sx }

        switch (type) {
          case 'text':
            questionComponents.push(
              <TextField {...commonProps} inputProps={{ maxLength }} rules={{ required: true }} />
            )
            break
          case 'select-one':
            questionComponents.push(
              <Select
                {...commonProps}
                options={inputOptions}
                emptyLabel={t('edit.step2.emptyLabel')}
                rules={{ required: true }}
              />
            )
            break
          case 'checkbox':
            questionComponents.push(
              <CheckboxGroup
                {...commonProps}
                options={inputOptions}
                rules={{
                  validate: (value) =>
                    (typeof value !== 'undefined' && value.length > 0) ||
                    t('edit.step2.multiCheckboxField.validation.mixed.required'),
                }}
              />
            )
            break
          case 'radio':
            questionComponents.push(
              <RadioGroup {...commonProps} options={inputOptions} rules={{ required: true }} />
            )
            break
          case 'switch':
            questionComponents.push(
              <RiskAnalysisSwitch
                {...commonProps}
                options={inputOptions}
                rules={{
                  validate: (value) =>
                    (typeof value === 'boolean' && value === true) ||
                    t('edit.step2.riskAnalysisSwitch.validation.boolean.isValue'),
                }}
              />
            )
            break
        }

        return questionComponents
      }

      for (let i = 0; i < questionIds.length; i++) {
        const questionId = questionIds[i]

        const question = questions[questionId] as QuestionV2

        const inputOptions = (question.options
          ?.map((option) => parseOption(option, question))
          .filter(identity) || []) as Array<InputOption>

        formComponents.push(
          ...buildFormQuestionComponents(question, inputOptions, questionIds.length - 1 === i)
        )
      }

      return { formComponents, isSubmitBtnDisabled: false }
    },
  },
}

export function getFormOperations(version: string) {
  const dynamicFormOperations = dynamicFormOperationsVersions[version]

  if (!dynamicFormOperations) {
    throw new Error('There is no business logic for this specific risk analysis data form')
  }

  return dynamicFormOperations
}
