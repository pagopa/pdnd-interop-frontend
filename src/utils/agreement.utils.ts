import type { AgreementSummary } from '@/types/agreement.types'
import type { ProviderOrConsumer } from '@/types/common.types'

export const canAgreementBeUpgraded = (agreement: AgreementSummary, mode: ProviderOrConsumer) => {
  if (!agreement || mode !== 'consumer') return false

  return !!(
    agreement.eservice.activeDescriptor &&
    agreement.eservice.activeDescriptor.state === 'PUBLISHED' &&
    agreement.eservice.activeDescriptor.version > agreement.eservice.version &&
    agreement.state !== 'ARCHIVED'
  )
}
