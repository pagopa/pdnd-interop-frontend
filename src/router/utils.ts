import { routes } from '@/router/routes'
import type { LangCode, ProviderOrConsumer } from '@/types/common.types'
import type { RouteConfig, RouteKey } from '@/router/types'
import { generatePath, matchPath } from 'react-router-dom'
import { getKeys } from '@/utils/array.utils'
import memoize from 'lodash/memoize'
import identity from 'lodash/identity'
import sortBy from 'lodash/sortBy'
import { PUBLIC_URL } from '@/config/env'
import qs from 'qs'
import QueryString from 'qs'

/** Returns the localized path of the given routeKey and language  */
export function getLocalizedRoutePathname(routeKey: RouteKey, lang: LangCode) {
  return `${PUBLIC_URL}/${lang}/${routes[routeKey].PATH[lang]}`
}

/** Checks if the routeKey represent the given pathname */
function matchRouteKeyPath(pathname: string, lang: LangCode, routeKey: RouteKey) {
  return matchPath(getLocalizedRoutePathname(routeKey as RouteKey, lang), pathname)
}

/** Returns the routeKey of the given pathname */
export const getRouteKeyFromPathname = memoize(function _getRouteKeyFromPathname(
  pathname: string,
  lang: LangCode
) {
  let pathnameWithBaseUrl = pathname
  if (!pathnameWithBaseUrl.startsWith(PUBLIC_URL)) {
    pathnameWithBaseUrl = PUBLIC_URL + pathnameWithBaseUrl
  }
  const isTheRouteKeyPath = matchRouteKeyPath.bind(null, pathnameWithBaseUrl, lang)
  const routeKey = getKeys(routes).find(isTheRouteKeyPath)

  if (!routeKey) {
    throw new Error(`Pathname ${pathnameWithBaseUrl} has no associated routeKey.`)
  }

  return routeKey
})

export function switchPathLang(fromLang: LangCode, toLang: LangCode) {
  for (const routeKey of Object.keys(routes)) {
    const match = matchRouteKeyPath(window.location.pathname, fromLang, routeKey as RouteKey)

    if (match) {
      let newPath = generatePath(
        getLocalizedRoutePathname(routeKey as RouteKey, toLang),
        match.params as Record<string, string>
      )

      if (window.location.search) {
        newPath += window.location.search
      }

      if (window.location.hash) {
        newPath += window.location.hash
      }

      window.location.replace(newPath)
      return
    }
  }
}

export const URL_FRAGMENTS: Record<string, Record<LangCode, string>> = {
  FIRST_DRAFT: { it: 'prima-bozza', en: 'first-draft' },
  EDIT: { it: 'modifica', en: 'bozza' },
}

export function getSplittedPath(route: RouteConfig, currentLang: LangCode) {
  return route.PATH[currentLang].split('/').filter(identity)
}

export const getParentRoutes = memoize((routeKey: RouteKey): Array<RouteKey> => {
  function isParentRoute(
    possibleParentRouteSubpaths: Array<string>,
    currentSubpaths: Array<string>
  ) {
    if (possibleParentRouteSubpaths.length >= currentSubpaths.length) {
      return false
    }

    const allSameFragments = possibleParentRouteSubpaths.every(
      (pathFragment, i) => pathFragment === currentSubpaths[i]
    )

    // URL_FRAGMENTS.EDIT is appended at the end of a read path of the same type.
    // E.g. /eservice/myid becomes /eservice/myid/edit. So it's always same length + 1
    const isFalseParent = currentSubpaths.length === possibleParentRouteSubpaths.length + 1

    if (allSameFragments && isEditPath(routeKey) && isFalseParent) {
      return false
    }

    return allSameFragments
  }

  const route = routes[routeKey]
  const currentSubpaths = getSplittedPath(route, 'it')

  const parents = Object.entries(routes).filter(([_, possibleParentRoute]) =>
    isParentRoute(getSplittedPath(possibleParentRoute, 'it'), currentSubpaths)
  )

  const sortedParents = sortBy(parents, ([_, parent]) => getSplittedPath(parent, 'it').length)

  return sortedParents.map(([routeKey]) => routeKey) as Array<RouteKey>
})

export const isProviderOrConsumerRoute = memoize(
  (routeKey: RouteKey): ProviderOrConsumer | null => {
    const excludeList = ['ui', 'it']
    const subroutes = getSplittedPath(routes[routeKey], 'it').filter(
      (b) => !excludeList.includes(b)
    )
    const [mode] = subroutes

    if (mode === 'erogazione') {
      return 'provider'
    }

    if (mode === 'fruizione') {
      return 'consumer'
    }

    return null
  }
)

export const isEditPath = memoize((routeKey: RouteKey): boolean => {
  const subroutes = getSplittedPath(routes[routeKey], 'it')

  const lastBit = subroutes[subroutes.length - 1]
  return Object.values(URL_FRAGMENTS.EDIT).some((f) => lastBit.endsWith(f))
})

export function parseSearch(search: string) {
  return qs.parse(search, { ignoreQueryPrefix: true })
}

export function stringifySearch(searchObj: Record<string, string> | QueryString.ParsedQs) {
  return qs.stringify(searchObj)
}

/**
 * Returns an array with all the dynamic path names for a given RouteKey
 * @example
 * "/:foo/test/:bar" => ["foo", "bar"]
 */
export const getDynamicPathsName = memoize((routeKey: RouteKey) => {
  return routes[routeKey].PATH.it
    .split('/')
    .filter(identity)
    .filter((subpath) => subpath.startsWith(':'))
    .map((param) => param.replace(':', ''))
})
