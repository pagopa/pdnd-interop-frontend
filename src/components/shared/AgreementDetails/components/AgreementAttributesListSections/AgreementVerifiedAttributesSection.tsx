import React from 'react'
import { useCurrentRoute } from '@/router'
import { useTranslation } from 'react-i18next'
import { useAgreementDetailsContext } from '../../AgreementDetailsContext'
import {
  SectionContainer,
  AttributeGroupContainer,
  AttributeContainer,
} from '@/components/layout/containers'
import { Stack } from '@mui/material'
import type { RemappedEServiceAttribute } from '@/types/attribute.types'
import {
  isAttributeOwned,
  isAttributeGroupFullfilled,
  isAttributeRevoked,
} from '@/utils/attribute.utils'
import type { ProviderOrConsumer } from '@/types/common.types'
import { useAgreementGetVerifiedAttributesActions } from '../../hooks/useAgreementGetVerifiedAttributesActions'

export const AgreementVerifiedAttributesSection: React.FC = () => {
  const { t: tAttribute } = useTranslation('attribute')
  const { mode } = useCurrentRoute()

  const { eserviceAttributes, partyAttributes } = useAgreementDetailsContext()

  const providerOrConsumer = mode as ProviderOrConsumer

  const verifiedAttributeGroups = eserviceAttributes?.verified ?? []
  const ownedVerifiedAttributes = partyAttributes?.verified ?? []

  const getAttributeActions = useAgreementGetVerifiedAttributesActions()

  const getChipLabel = (attributeId: string) => {
    const attribute = ownedVerifiedAttributes.find((a) => a.id === attributeId)
    if (!attribute) return

    if (mode === 'consumer' && isAttributeRevoked('verified', attribute))
      return tAttribute('group.manage.revokedByParty')
  }

  function getGroupContainerProps(
    group: RemappedEServiceAttribute
  ): React.ComponentProps<typeof AttributeGroupContainer> {
    const isGroupFulfilled = isAttributeGroupFullfilled('verified', ownedVerifiedAttributes, group)

    if (isGroupFulfilled) {
      return {
        title: tAttribute(`group.manage.success.${providerOrConsumer}`),
        color: 'success',
      }
    }

    return {
      title: tAttribute(`group.manage.warning.verified.${providerOrConsumer}`),
      color: 'warning',
    }
  }

  return (
    <SectionContainer
      newDesign
      innerSection
      title={tAttribute('verified.label')}
      description={tAttribute('verified.description')}
    >
      <Stack spacing={2}>
        {verifiedAttributeGroups.map((group, i) => (
          <AttributeGroupContainer {...getGroupContainerProps(group)} key={i}>
            <Stack spacing={1.2} sx={{ my: 2, mx: 0, listStyle: 'none', px: 0 }} component="ul">
              {group.attributes.map((attribute) => (
                <AttributeContainer
                  key={attribute.id}
                  attribute={attribute}
                  chipLabel={getChipLabel(attribute.id)}
                  checked={isAttributeOwned('verified', attribute.id, ownedVerifiedAttributes)}
                  actions={getAttributeActions(attribute.id)}
                />
              ))}
            </Stack>
          </AttributeGroupContainer>
        ))}
      </Stack>
      {verifiedAttributeGroups.length === 0 && (
        <AttributeGroupContainer
          title={tAttribute(`noAttributesRequiredAlert.${providerOrConsumer}`, {
            attributeKey: tAttribute(`type.verified_other`),
          })}
          color="gray"
        />
      )}
    </SectionContainer>
  )
}
