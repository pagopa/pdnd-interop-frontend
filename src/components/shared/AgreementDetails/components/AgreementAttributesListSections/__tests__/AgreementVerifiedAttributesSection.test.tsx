import React from 'react'
import { mockUseCurrentRoute, renderWithApplicationContext } from '@/utils/testing.utils'
import { mockAgreementDetailsContext } from '../../__tests__/test.commons'
import { AgreementVerifiedAttributesSection } from '../AgreementVerifiedAttributesSection'
import {
  createVerifiedTenantAttribute,
  createMockRemappedEServiceAttribute,
} from '__mocks__/data/attribute.mocks'

mockUseCurrentRoute({ mode: 'provider' })

describe('AgreementVerifiedAttributesSection', () => {
  it('should match snapshot when e-service does not require verified attributes', () => {
    mockAgreementDetailsContext({
      eserviceAttributes: undefined,
    })
    const { baseElement } = renderWithApplicationContext(<AgreementVerifiedAttributesSection />, {
      withReactQueryContext: true,
    })
    expect(baseElement).toMatchSnapshot()
  })

  it('should match snapshot when e-service requires verified attributes', () => {
    mockAgreementDetailsContext({
      eserviceAttributes: {
        certified: [],
        verified: [
          createMockRemappedEServiceAttribute({ attributes: [{ id: 'a-1-1' }, { id: 'a-1-2' }] }),
          createMockRemappedEServiceAttribute({ attributes: [{ id: 'a-2-1' }, { id: 'a-2-2' }] }),
        ],
        declared: [],
      },
      partyAttributes: {
        certified: [],
        verified: [
          createVerifiedTenantAttribute({
            id: 'a-1-1',
            revokedBy: [],
            verifiedBy: [{ id: 'test' }],
          }),
        ],
        declared: [],
      },
    })
    const { baseElement } = renderWithApplicationContext(<AgreementVerifiedAttributesSection />, {
      withReactQueryContext: true,
    })
    expect(baseElement).toMatchSnapshot()
  })

  it('should show revoked chip when attribute is revoked and user is consumer', () => {
    mockUseCurrentRoute({ mode: 'consumer' })
    mockAgreementDetailsContext({
      eserviceAttributes: {
        certified: [],
        verified: [createMockRemappedEServiceAttribute({ attributes: [{ id: 'a-1-1' }] })],
        declared: [],
      },
      partyAttributes: {
        certified: [],
        verified: [
          createVerifiedTenantAttribute({
            id: 'a-1-1',
            revokedBy: [{ id: 'test' }],
            verifiedBy: [],
          }),
        ],
        declared: [],
      },
    })

    const screen = renderWithApplicationContext(<AgreementVerifiedAttributesSection />, {
      withReactQueryContext: true,
    })

    expect(screen.getByText('group.manage.revokedByProducer')).toBeInTheDocument()
  })
})
