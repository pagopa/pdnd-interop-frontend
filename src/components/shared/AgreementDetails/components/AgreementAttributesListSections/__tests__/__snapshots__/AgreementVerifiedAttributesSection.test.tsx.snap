// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AgreementVerifiedAttributesSection > should match snapshot when e-service does not require verified attributes 1`] = `
<body>
  <div />
</body>
`;

exports[`AgreementVerifiedAttributesSection > should match snapshot when e-service requires verified attributes 1`] = `
<body>
  <div>
    <section
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1od456z-MuiPaper-root"
    >
      <div
        class="MuiStack-root css-jfdv4h-MuiStack-root"
      >
        <div
          class="MuiStack-root css-3y1137-MuiStack-root"
        >
          <h3
            class="MuiTypography-root MuiTypography-sidenav css-37jdci-MuiTypography-root"
          >
            verified.label
          </h3>
        </div>
        <p
          class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
        >
          verified.description
        </p>
      </div>
      <div
        class="MuiBox-root css-1yuhvjn"
      >
        <div
          class="MuiStack-root css-1sazv7p-MuiStack-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bj42sx-MuiPaper-root-MuiCard-root"
          >
            <div
              class="MuiCardHeader-root css-1g82uv9-MuiCardHeader-root"
            >
              <div
                class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
              >
                <span
                  class="MuiTypography-root MuiTypography-body1 MuiCardHeader-title css-153kzzj-MuiTypography-root"
                >
                  group.manage.success.provider
                </span>
              </div>
            </div>
            <div
              class="MuiCardContent-root css-1jqdwvk-MuiCardContent-root"
            >
              <ul
                class="MuiStack-root css-slre3t-MuiStack-root"
              >
                <div
                  class="MuiStack-root css-1sos3zc-MuiStack-root"
                >
                  <div
                    class="MuiStack-root css-1sos3zc-MuiStack-root"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-4b67si-MuiSvgIcon-root"
                      data-testid="CheckCircleIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                      />
                    </svg>
                  </div>
                  <div
                    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-xr63ld-MuiPaper-root-MuiCard-root"
                  >
                    <div
                      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAccordion-root MuiAccordion-rounded css-1byhwgv-MuiPaper-root-MuiAccordion-root"
                    >
                      <div
                        aria-controls=":r1:"
                        aria-expanded="false"
                        class="MuiButtonBase-root MuiAccordionSummary-root css-1pvvkxv-MuiButtonBase-root-MuiAccordionSummary-root"
                        id=":r2:"
                        role="button"
                        tabindex="0"
                      >
                        <div
                          class="MuiAccordionSummary-content css-1betqn-MuiAccordionSummary-content"
                        >
                          <p
                            class="MuiTypography-root MuiTypography-body1 css-1qqynwt-MuiTypography-root"
                          >
                            Attribute Name
                          </p>
                        </div>
                        <div
                          class="MuiAccordionSummary-expandIconWrapper css-yw020d-MuiAccordionSummary-expandIconWrapper"
                        >
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                            data-testid="ExpandMoreIcon"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
                            />
                          </svg>
                        </div>
                      </div>
                      <div
                        class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-bz4dnt-MuiCollapse-root"
                        style="min-height: 0px;"
                      >
                        <div
                          class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
                        >
                          <div
                            class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
                          >
                            <div
                              aria-labelledby=":r2:"
                              class="MuiAccordion-region"
                              id=":r1:"
                              role="region"
                            >
                              <div
                                class="MuiAccordionDetails-root css-15v22id-MuiAccordionDetails-root"
                              />
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="MuiStack-root css-1sos3zc-MuiStack-root"
                >
                  <div
                    class="MuiStack-root css-1sos3zc-MuiStack-root"
                  />
                  <div
                    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-xr63ld-MuiPaper-root-MuiCard-root"
                  >
                    <div
                      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAccordion-root MuiAccordion-rounded css-1byhwgv-MuiPaper-root-MuiAccordion-root"
                    >
                      <div
                        aria-controls=":r3:"
                        aria-expanded="false"
                        class="MuiButtonBase-root MuiAccordionSummary-root css-1pvvkxv-MuiButtonBase-root-MuiAccordionSummary-root"
                        id=":r4:"
                        role="button"
                        tabindex="0"
                      >
                        <div
                          class="MuiAccordionSummary-content css-1betqn-MuiAccordionSummary-content"
                        >
                          <p
                            class="MuiTypography-root MuiTypography-body1 css-1qqynwt-MuiTypography-root"
                          >
                            Attribute Name
                          </p>
                        </div>
                        <div
                          class="MuiAccordionSummary-expandIconWrapper css-yw020d-MuiAccordionSummary-expandIconWrapper"
                        >
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                            data-testid="ExpandMoreIcon"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
                            />
                          </svg>
                        </div>
                      </div>
                      <div
                        class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-bz4dnt-MuiCollapse-root"
                        style="min-height: 0px;"
                      >
                        <div
                          class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
                        >
                          <div
                            class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
                          >
                            <div
                              aria-labelledby=":r4:"
                              class="MuiAccordion-region"
                              id=":r3:"
                              role="region"
                            >
                              <div
                                class="MuiAccordionDetails-root css-15v22id-MuiAccordionDetails-root"
                              />
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </ul>
            </div>
          </div>
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-dxqih4-MuiPaper-root-MuiCard-root"
          >
            <div
              class="MuiCardHeader-root css-97snmn-MuiCardHeader-root"
            >
              <div
                class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
              >
                <span
                  class="MuiTypography-root MuiTypography-body1 MuiCardHeader-title css-153kzzj-MuiTypography-root"
                >
                  group.manage.warning.verified.provider
                </span>
              </div>
            </div>
            <div
              class="MuiCardContent-root css-1jqdwvk-MuiCardContent-root"
            >
              <ul
                class="MuiStack-root css-slre3t-MuiStack-root"
              >
                <div
                  class="MuiStack-root css-1sos3zc-MuiStack-root"
                >
                  <div
                    class="MuiStack-root css-1sos3zc-MuiStack-root"
                  />
                  <div
                    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-xr63ld-MuiPaper-root-MuiCard-root"
                  >
                    <div
                      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAccordion-root MuiAccordion-rounded css-1byhwgv-MuiPaper-root-MuiAccordion-root"
                    >
                      <div
                        aria-controls=":r5:"
                        aria-expanded="false"
                        class="MuiButtonBase-root MuiAccordionSummary-root css-1pvvkxv-MuiButtonBase-root-MuiAccordionSummary-root"
                        id=":r6:"
                        role="button"
                        tabindex="0"
                      >
                        <div
                          class="MuiAccordionSummary-content css-1betqn-MuiAccordionSummary-content"
                        >
                          <p
                            class="MuiTypography-root MuiTypography-body1 css-1qqynwt-MuiTypography-root"
                          >
                            Attribute Name
                          </p>
                        </div>
                        <div
                          class="MuiAccordionSummary-expandIconWrapper css-yw020d-MuiAccordionSummary-expandIconWrapper"
                        >
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                            data-testid="ExpandMoreIcon"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
                            />
                          </svg>
                        </div>
                      </div>
                      <div
                        class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-bz4dnt-MuiCollapse-root"
                        style="min-height: 0px;"
                      >
                        <div
                          class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
                        >
                          <div
                            class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
                          >
                            <div
                              aria-labelledby=":r6:"
                              class="MuiAccordion-region"
                              id=":r5:"
                              role="region"
                            >
                              <div
                                class="MuiAccordionDetails-root css-15v22id-MuiAccordionDetails-root"
                              />
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="MuiStack-root css-1sos3zc-MuiStack-root"
                >
                  <div
                    class="MuiStack-root css-1sos3zc-MuiStack-root"
                  />
                  <div
                    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-xr63ld-MuiPaper-root-MuiCard-root"
                  >
                    <div
                      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAccordion-root MuiAccordion-rounded css-1byhwgv-MuiPaper-root-MuiAccordion-root"
                    >
                      <div
                        aria-controls=":r7:"
                        aria-expanded="false"
                        class="MuiButtonBase-root MuiAccordionSummary-root css-1pvvkxv-MuiButtonBase-root-MuiAccordionSummary-root"
                        id=":r8:"
                        role="button"
                        tabindex="0"
                      >
                        <div
                          class="MuiAccordionSummary-content css-1betqn-MuiAccordionSummary-content"
                        >
                          <p
                            class="MuiTypography-root MuiTypography-body1 css-1qqynwt-MuiTypography-root"
                          >
                            Attribute Name
                          </p>
                        </div>
                        <div
                          class="MuiAccordionSummary-expandIconWrapper css-yw020d-MuiAccordionSummary-expandIconWrapper"
                        >
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                            data-testid="ExpandMoreIcon"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
                            />
                          </svg>
                        </div>
                      </div>
                      <div
                        class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-bz4dnt-MuiCollapse-root"
                        style="min-height: 0px;"
                      >
                        <div
                          class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
                        >
                          <div
                            class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
                          >
                            <div
                              aria-labelledby=":r8:"
                              class="MuiAccordion-region"
                              id=":r7:"
                              role="region"
                            >
                              <div
                                class="MuiAccordionDetails-root css-15v22id-MuiAccordionDetails-root"
                              />
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </section>
  </div>
</body>
`;
